UPGRADE.md

```diff
@@ -1411,6 +1411,71 @@ const Layout = (props) => {
 };
 ```
 
+## Form Submissions and Side Effects are Easier
+
+We previously had a complex solution for having multiple submit buttons: a SaveContext providing side effects modifiers and refs to the current ones, and also a redirect prop. However, this was redundant and confusing as the `save` function provided by our mutation hooks also accepted side effect override at call time. It was even more confusing when adding custom or multiple `<SaveButton>` as those received two submission related functions: `handleSubmit` and `handleSubmitWithRedirect`.
+
+Besides, our solution prevented the native browser submit on enter feature and this was an accessibility issue for some users such as Japanese people.
+
+The new solution leverage the fact that we already make the `save` function available through context (`useSaveContext`) and users can get the current form values, call the save function directly and override any of the side effects as they usually do with mutation hooks.
+
+We also added side effects props on the SaveButton to make most common cases possible and avoid gotchas such as preventing default event.
+
+If you relied on `handleSubmit` or `handleSubmitWithRedirect`, you can now use the `SaveButton` and override any of the side effect props: `onSuccess`, `onFailure` or `transform`.
+
+The `FormWithRedirect`, `SimpleForm`, `TabbedForm` and `SaveButton` don't have a `redirect` prop anymore. You should handle redirection either:
+- in the submit handler if you provided a custom one, leveraging the `useRedirect` hook
+- at the `Create` or `Edit` level as they now accept a `redirect` prop
+- in the `SaveButton` side effects props (`onSuccess` or `onFailure`), leveraging the `useRedirect` hook
+
+If you had the `redirect` prop set on the form component, move it to `Create` or `Edit` component:
+
+```diff
+-<Create>
++<Create redirect="edit">
+-    <SimpleForm redirect="edit">
++    <SimpleForm>
+        <TextInput source="name">
+    </SimpleForm>
+</Create>
+```
+
+If you had the `redirect` prop set on the `SaveButton`, provide a `onSuccess` prop:
+
+```diff
+const PostCreateToolbar = props => {
++    const redirect = useRedirect();    
+    return (
+        <Toolbar {...props}>
+            <SaveButton
+                label="post.action.save_and_edit"
+-                redirect="edit"
++                onSuccess={data => {
++                    redirect('edit', '/posts', data.id)
```

Comment:
this is missing the default notification

Reply:
Added